// nbt-parser/dist/serialization/snbt/serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeTagToSNBT = serializeTagToSNBT;
exports.serializePayloadToSNBT = serializePayloadToSNBT;
const Tag_1 = require("../../nbt/Tag");
const types_1 = require("../../types");
const Payload_1 = require("../../nbt/Payload");
const utils_1 = require("./utils");
const defaultOptions = {
    format: "pretty",
    preferUnquotedString: true,
    quote: "prefer-double",
    tab: "    ",
    preferBoolean: false,
    breakLine: 33,
};
function serializeTagToSNBT(tag, options) {
    const fullOptions = {
        ...defaultOptions,
        ...options
    };
    if (tag.root)
        return serializePayloadToSNBT(tag.payload, fullOptions);
    return serializePayloadToSNBT(new Payload_1.CompoundPayload([tag]), fullOptions);
}
function serializeTagToUnwrappedSNBT(tag, options) {
    const space = options.format == "compressed" ? "" : " ";
    const name = canUseUnquotedString(tag.name, options) ? tag.name : serializeString(tag.name, options);
    return `${name}:${space}${serializePayloadToSNBT(tag.payload, options)}`;
}
function serializePayloadToSNBT(payload, options) {
    const fullOptions = {
        ...defaultOptions,
        ...options
    };
    switch (payload.tagId) {
        case Tag_1.TagId.BYTE:
            if (fullOptions.preferBoolean && payload.value == 0)
                return "false";
            if (fullOptions.preferBoolean && payload.value == 1)
                return "true";
            return `${payload.value}b`;
        case Tag_1.TagId.SHORT:
            return `${payload.value}s`;
        case Tag_1.TagId.INT:
            return `${payload.value}`;
        case Tag_1.TagId.LONG:
            return `${payload.value}l`;
        case Tag_1.TagId.FLOAT:
            return `${payload.value}f`;
        case Tag_1.TagId.DOUBLE:
            return `${payload.value}d`;
        case Tag_1.TagId.BYTE_ARRAY:
            return serializeListPayloadToSNBT(payload, fullOptions, "B;");
        case Tag_1.TagId.STRING:
            return serializeString(payload.value, fullOptions);
        case Tag_1.TagId.LIST:
            return serializeListPayloadToSNBT(payload, fullOptions, "");
        case Tag_1.TagId.COMPOUND:
            return serializeCompoundPayloadToSNBT(payload, fullOptions);
        case Tag_1.TagId.INT_ARRAY:
            return serializeListPayloadToSNBT(payload, fullOptions, "I;");
        case Tag_1.TagId.LONG_ARRAY:
            return serializeListPayloadToSNBT(payload, fullOptions, "L;");
        default:
            throw new types_1.NBTError(`Unknown tag id: ${payload.tagId}`);
    }
}
function canUseUnquotedString(string, options) {
    return options.preferUnquotedString && /^[a-zA-Z_.][a-zA-Z_.\d+-]*$/.test(string);
}
function serializeString(string, options) {
    let quote;
    switch (options.quote) {
        case "prefer-double":
            quote = (string.includes("'") && !string.includes("\"")) ? "'" : "\"";
            break;
        case "prefer-single":
            quote = (string.includes("\"") && !string.includes("'")) ? "\"" : "'";
            break;
        case "double":
            quote = "\"";
            break;
        case "single":
            quote = "'";
            break;
        default:
            throw new types_1.NBTError(`Unknown quote preference: ${options.quote}`);
    }
    return `${quote}${(0, utils_1.escapeString)(string, quote)}${quote}`;
}
function serializeCompoundPayloadToSNBT(payload, options) {
    switch (options.format) {
        case "multiline": {
            const pretty = serializeCompoundPayloadToSNBT(payload, {
                ...options,
                format: "pretty"
            });
            if (pretty.length < options.breakLine)
                return pretty;
            return `{\n${options.tab}${payload.value
                .map((tag) => serializeTagToUnwrappedSNBT(tag, options))
                .join(",\n")
                .replaceAll("\n", `\n${options.tab}`)}\n}`;
        }
        case "pretty":
            return `{ ${payload.value.map((tag) => serializeTagToUnwrappedSNBT(tag, options)).join(", ")} }`;
        case "compressed":
            return `{${payload.value.map((tag) => serializeTagToUnwrappedSNBT(tag, options)).join(",")}}`;
        default:
            throw new types_1.NBTError(`Unknown compression: ${options.format}`);
    }
}
function serializeListPayloadToSNBT(payload, options, prefix) {
    switch (options.format) {
        case "multiline": {
            const pretty = serializeListPayloadToSNBT(payload, {
                ...options,
                format: "pretty"
            }, prefix);
            if (pretty.length < options.breakLine)
                return pretty;
            return `[${prefix}\n${options.tab}${payload.value
                .map((p) => serializePayloadToSNBT(p, options))
                .join(",\n")
                .replaceAll("\n", `\n${options.tab}`)}\n]`;
        }
        case "pretty":
            return `[${prefix} ${payload.value.map((p) => serializePayloadToSNBT(p, options)).join(", ")} ]`;
        case "compressed":
            return `[${prefix}${payload.value.map((p) => serializePayloadToSNBT(p, options)).join(",")}]`;
        default:
            throw new types_1.NBTError(`Unknown compression: ${options.format}`);
    }
}
