// nbt-parser/dist/nbt/Payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LongArrayPayload = exports.IntArrayPayload = exports.CompoundPayload = exports.ListPayload = exports.StringPayload = exports.ByteArrayPayload = exports.DoublePayload = exports.FloatPayload = exports.LongPayload = exports.IntPayload = exports.ShortPayload = exports.BytePayload = exports.AbstractPayload = void 0;
const types_1 = require("../types");
const Tag_1 = require("./Tag");
const serializer_1 = require("../serialization/nbt/serializer");
const serializer_2 = require("../serialization/json/serializer");
const serializer_3 = require("../serialization/snbt/serializer");
class AbstractPayload {
    constructor(value) {
        this._value = this.validate(value) ?? value;
    }
    validate(value) {
    }
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = this.validate(value) ?? value;
    }
    copy(value) {
        return new this.constructor(value ?? this.value);
    }
    equals(other) {
        return this.tagId == other.tagId && this.value == other.value;
    }
    toNBT(edition) {
        return (0, serializer_1.serializePayloadToNBT)(this, edition);
    }
    toJson() {
        return (0, serializer_2.serializePayloadToJson)(this);
    }
    toSNBT(compression) {
        return (0, serializer_3.serializePayloadToSNBT)(this, compression);
    }
}
exports.AbstractPayload = AbstractPayload;
class BytePayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        return new Int8Array([value])[0];
    }
    get tagId() {
        return Tag_1.TagId.BYTE;
    }
}
exports.BytePayload = BytePayload;
class ShortPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        return new Int16Array([value])[0];
    }
    get tagId() {
        return Tag_1.TagId.SHORT;
    }
}
exports.ShortPayload = ShortPayload;
class IntPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        return new Int32Array([value])[0];
    }
    get tagId() {
        return Tag_1.TagId.INT;
    }
}
exports.IntPayload = IntPayload;
class LongPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        return new BigInt64Array([value])[0];
    }
    get tagId() {
        return Tag_1.TagId.LONG;
    }
}
exports.LongPayload = LongPayload;
class FloatPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (!Number.isFinite(value) || Number.isNaN(value))
            throw new types_1.NBTError(`Invalid number value: ${value}`);
        return new Float32Array([value])[0];
    }
    get tagId() {
        return Tag_1.TagId.FLOAT;
    }
}
exports.FloatPayload = FloatPayload;
class DoublePayload extends AbstractPayload {
    validate(value) {
        if (!Number.isFinite(value) || Number.isNaN(value))
            throw new types_1.NBTError(`Invalid number value: ${value}`);
    }
    constructor(value) {
        super(value);
    }
    get tagId() {
        return Tag_1.TagId.DOUBLE;
    }
}
exports.DoublePayload = DoublePayload;
class ByteArrayPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (value.length >= 2 ** 31)
            throw new types_1.NBTError(`Byte array is longer than 2147483647 items: [${value.map(p => p.toJson()).join(", ")}]`);
    }
    get tagId() {
        return Tag_1.TagId.BYTE_ARRAY;
    }
}
exports.ByteArrayPayload = ByteArrayPayload;
class StringPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (value.length > 65535)
            throw new types_1.NBTError(`String is longer than 65535 chars: ${value}`);
    }
    get tagId() {
        return Tag_1.TagId.STRING;
    }
}
exports.StringPayload = StringPayload;
class ListPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (value.length >= 2 ** 31)
            throw new types_1.NBTError(`List is longer than 65535 items: [${value.map(p => p.toJson()).join()}]`);
        if (value.length > 0) {
            const firstTag = value[0].tagId;
            if (value.some(payload => payload.tagId != firstTag))
                throw new types_1.NBTError(`List contains different types, tag ids: [${value.reduce((acc, cur) => {
                    if (!acc.includes(cur.tagId))
                        acc.push(cur.tagId);
                    return acc;
                }, []).sort().join(", ")}]`);
        }
    }
    get tagId() {
        return Tag_1.TagId.LIST;
    }
}
exports.ListPayload = ListPayload;
class CompoundPayload extends AbstractPayload {
    // TODO: throw error when depth > 512
    constructor(value) {
        super(value);
    }
    get tagId() {
        return Tag_1.TagId.COMPOUND;
    }
}
exports.CompoundPayload = CompoundPayload;
class IntArrayPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (value.length >= 2 ** 31)
            throw new types_1.NBTError(`Int array is longer than 2147483647 items: [${value.map(p => p.toJson()).join(", ")}]`);
    }
    get tagId() {
        return Tag_1.TagId.INT_ARRAY;
    }
}
exports.IntArrayPayload = IntArrayPayload;
class LongArrayPayload extends AbstractPayload {
    constructor(value) {
        super(value);
    }
    validate(value) {
        if (value.length >= 2 ** 31)
            throw new types_1.NBTError(`Long array is longer than 2147483647 items: [${value.map(p => p.toJson()).join(", ")}]`);
    }
    get tagId() {
        return Tag_1.TagId.LONG_ARRAY;
    }
}
exports.LongArrayPayload = LongArrayPayload;
