// nbt-parser/dist/nbt/Tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tag = exports.TagId = void 0;
const types_1 = require("../types");
const serializer_1 = require("../serialization/nbt/serializer");
const serializer_2 = require("../serialization/json/serializer");
const serializer_3 = require("../serialization/snbt/serializer");
var TagId;
(function (TagId) {
    TagId[TagId["END"] = 0] = "END";
    TagId[TagId["BYTE"] = 1] = "BYTE";
    TagId[TagId["SHORT"] = 2] = "SHORT";
    TagId[TagId["INT"] = 3] = "INT";
    TagId[TagId["LONG"] = 4] = "LONG";
    TagId[TagId["FLOAT"] = 5] = "FLOAT";
    TagId[TagId["DOUBLE"] = 6] = "DOUBLE";
    TagId[TagId["BYTE_ARRAY"] = 7] = "BYTE_ARRAY";
    TagId[TagId["STRING"] = 8] = "STRING";
    TagId[TagId["LIST"] = 9] = "LIST";
    TagId[TagId["COMPOUND"] = 10] = "COMPOUND";
    TagId[TagId["INT_ARRAY"] = 11] = "INT_ARRAY";
    TagId[TagId["LONG_ARRAY"] = 12] = "LONG_ARRAY";
})(TagId || (exports.TagId = TagId = {}));
class Tag {
    constructor(name, payload, root = false) {
        this.payload = payload;
        this.root = root;
        this.validate(name);
        this._name = name;
    }
    validate(name) {
        if (name.length > 65535)
            throw new types_1.NBTError(`Tag name is longer than 65535 chars: ${name}`);
        if (this.root && this.payload.tagId != TagId.COMPOUND)
            throw new types_1.NBTError(`Root tag must be a compound tag`);
    }
    copy(name, payload, root) {
        return new Tag(name ?? this._name, payload ?? this.payload.copy(), root ?? this.root);
    }
    ;
    equals(other) {
        return this._name == other._name && this.payload.equals(other.payload);
    }
    get name() {
        return this._name;
    }
    set name(name) {
        this.validate(name);
        this._name = name;
    }
    toNBT(edition, binaryCompression) {
        return (0, serializer_1.serializeTagToNBT)(this, edition, binaryCompression);
    }
    toJson() {
        if (this.root)
            return this.payload.toJson();
        else
            return (0, serializer_2.serializeTagToJson)(this);
    }
    toSNBT(compression) {
        if (this.root)
            return this.payload.toSNBT(compression);
        else
            return (0, serializer_3.serializeTagToSNBT)(this, compression);
    }
}
exports.Tag = Tag;
