// nbt-parser/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SNBTStringReader = void 0;
exports.concatUint8Arrays = concatUint8Arrays;
exports.isLittleEndian = isLittleEndian;
exports.uuidToIntArray = uuidToIntArray;
exports.intArrayToUuid = intArrayToUuid;
const types_1 = require("./types");
function concatUint8Arrays(...arrays) {
    let totalLength = 0;
    for (let arr of arrays) {
        totalLength += arr.length;
    }
    let result = new Uint8Array(totalLength);
    let offset = 0;
    for (let arr of arrays) {
        result.set(arr, offset);
        offset += arr.length;
    }
    return result;
}
function isLittleEndian(edition) {
    return edition === "bedrock";
}
function uuidToIntArray(uuid) {
    const hex = uuid.replace(/-/g, "").toLowerCase();
    const parts = [
        hex.substring(0, 8),
        hex.substring(8, 16),
        hex.substring(16, 24),
        hex.substring(24, 32),
    ];
    return parts.map(part => {
        const unsigned = parseInt(part, 16);
        return unsigned > 0x7FFFFFFF ? unsigned - 0x100000000 : unsigned;
    });
}
function intArrayToUuid(intArray) {
    const parts = intArray.map(part => {
        const unsigned = part >= 0 ? part : part + 0x100000000;
        return unsigned.toString(16).padStart(8, "0");
    });
    const slashBetween = (part) => `${part.slice(0, 4)}-${part.slice(4, 8)}`;
    return `${parts[0]}-${slashBetween(parts[1])}-${slashBetween(parts[2])}${parts[3]}`;
}
class SNBTStringReader {
    /**
     * Creates a new StringReader
     * @param string The string to read
     */
    constructor(string) {
        this.string = string;
        this._index = -1;
    }
    /**
     * Gets pointer position
     * @returns pointer position
     */
    get index() {
        return this._index;
    }
    /**
     * Sets pointer position
     * @param index position
     * @throws Error if position is out of bounds
     */
    set index(index) {
        if (index < -1 || index >= this.length)
            throw new Error(`Index ${index} out of string "${this.string}"`);
        this._index = index;
    }
    /**
     * Gets the length of the string
     *
     * Equals to <code>StringReader.string.length</code>
     * @returns the length of the string
     */
    get length() {
        return this.string.length;
    }
    /**
     * Reads the character at the specified index
     * @param index the position of the character to get, defaults to the next character
     * @throws Error if position is out of bounds
     * @returns the character at the specified index
     */
    charAt(index) {
        if (index < 0 || index >= this.length) {
            throw new Error(`Index ${index} out of string "${this.string}"`);
        }
        return this.string[index];
    }
    /**
     * Reads the next character
     *
     * Equals to <code>StringReader.charAt(StringReader.index + 1)</code>
     * @throws Error if the next position is out of bounds
     * @returns the character at the next position
     */
    peek() {
        return this.charAt(this.index + 1);
    }
    /**
     * Reads all characters starting from the current position
     *
     * Does not read the current character
     * @returns all characters starting from the current position
     */
    peekNext() {
        return this.string.slice(this.index + 1);
    }
    /**
     * Reads the current character
     *
     * Equals to <code>StringReader.charAt(StringReader.index)</code>
     * @returns current character
     */
    curr() {
        return this.charAt(this.index);
    }
    /**
     * Reads the previous character
     *
     * Equals to <code>StringReader.charAt(StringReader.index - 1)</code>
     * @throws Error if the previous position is out of bounds
     */
    peekPrev() {
        return this.charAt(this.index - 1);
    }
    /**
     * Moves the pointer to the previous character and returns it
     * @throws Error if the previous position is out of bounds
     */
    prev() {
        this.index--;
        return this.curr();
    }
    /**
     * Moves the pointer to the next character and returns it
     * @throws Error if the next position is out of bounds
     */
    next() {
        this.index++;
        return this.curr();
    }
    /**
     * Reads all characters until the specified test function returns false
     *
     * Does not read the current character
     *
     * Positions the pointer to the last read character
     * @param prediction the test function
     * @returns the read characters
     */
    nextUnless(prediction) {
        let str = "";
        while (!this.eof() && prediction(this.peek(), this)) {
            str += this.next();
        }
        return str;
    }
    /**
     * Moves the pointer to the specified position and returns it
     * @param pos the position to move to
     * @throws Error if position is out of bounds
     */
    moveTo(pos) {
        this.index = pos;
        return this.curr();
    }
    /**
     * Checks if the current position is at the end of the string
     * @returns true if the current position is at the end of the string, false otherwise
     */
    eof() {
        return this._index == this.length - 1;
    }
    /**
     * Skips all whitespace characters starting from the current position
     * and position the pointer before the first non-whitespace character
     */
    skipWhitespace() {
        this.nextUnless(char => /\s/.test(char));
    }
    /**
     * Expects the characters at the current position
     * @param chars the characters to expect
     * @throws Error if the position is out of bounds
     * @throws Error if the character is not one of the specified characters
     * @returns the expected character
     */
    expect(...chars) {
        const char = this.curr();
        if (!chars.includes(char))
            throw this.newSNBTError(`Expected character ${chars.map(c => `"${c}"`).join(" or ")} at index ${this.index}, but got "${char}"`);
        return char;
    }
    /**
     * Expects the characters at the next position and moves the pointer to the next position
     * @param chars the characters to expect
     * @throws Error if the position is out of bounds
     * @throws Error if the character is not one of the specified characters
     * @returns the expected character
     */
    expectNext(...chars) {
        this.next();
        return this.expect(...chars);
    }
    newSNBTError(message) {
        return new types_1.SNBTError(message, this.string, this.index);
    }
}
exports.SNBTStringReader = SNBTStringReader;
